docs/spec_draft.md

```diff
@@ -339,6 +349,170 @@ only exists to establish data dependencies from `result` to `inputs`.
 
 [Back to Ops](#index-of-ops)
 
+## stablehlo.all_to_all
+
+### Semantics
+
+Scatters `split_count` blocks of data, by splitting `operand` along
+`split_dimension`, from all the execution instances to all the execution
+instances. Each execution instance concatenates the received blocks along the
+`concat_dimension`.
+
+The following diagram shows how the scattered blocks of `operand`, from all the
+execution instances, are concatenated to form `result` in each execution
+instance.
+
+<img align="center" src="spec_draft/all_to_all.svg" />
+
+In the instruction syntax, multiple replica ids are grouped as a replica
+subgroup and multiple replica subgroups form `replica_groups`.
+
+The execution of the operation involves logically partitioning each execution
+instance (`ei`) of `rid`, `pid` pair, henceforth referred to as `ei(rid, pid)`
+into groups. All the execution instances within a group can participate in the
+operation without getting interfered by execution instances from other groups.
+The number of groups represents the number of `stablehlo.all_to_all` operations
+executed in parallel.
+
+Given,
+  * `num_pids`: Number of partition ids
+  * `num_rids`: Number of replica ids
+  * `all_partions_ids`: { pid : pid $\in$ [0, `num_pids`) }
+  * `all_replica_ids`: { rid : rid $\in$ [0, `num_rids`) }
+  * `size_subgroup`: size of each subgroup in `replica_groups`.
+  * `num_subgroup`: number of subgroups in `replica_groups`.
+
+The formation of groups is defined as follows:
+   * Number of groups formed: `num_subgroup` * `num_pids`.
+
+   * Size of each group: `size_subgroup`.
+
+   * The formation of `groups` can be demonstrated, using Python-like syntax, as
+     follows:
+     ```python
+     all_replica_ids = [rid for rid in range(num_rids)]
+     all_partition_ids = [pid for pid in range(num_pids)]
+     groups = []
+
+     if len(replica_groups) == 0:
+       replica_groups = [all_replica_ids]
+
+     for replica_group in replica_groups:
+       for pid in all_partition_ids:
+         sub_group = []
+         for rid in replica_group:
+           sub_group.append(ei(rid, pid))
+         groups.append(sub_group)
+     return groups
+     ```
+
+  * For example, assuming `num_pids = 2` and
+      `replica_groups = [[0, 2], [1, 3]]`:
+      `groups` formed: `[ei(0, 0), ei(2, 0)], [ei(0, 1), ei(2, 1)], [ei(1, 0), ei(3, 0)], and [ei(1, 1), ei(3, 1)]`.
+
+For each group `G` $\in$ `groups`, the operation can be described as:
+  * Each `operand`, corresponding to an execution instance in `G`,
+    is split into `split_count` number of blocks, described below using
+    Python-like syntax.
+    ```python
```

Comment:
Feedback:
* It's probably wise to use `//` for integer division here and below:
* Few minor typos
```python
blocks = [
    slice(
      operand=operand,
      start_indices=[s0, s1, ..., sR-1],
        # where
        #  - sk = 0 if k != split_dimension
        #  - sk = j * dim(operand) // split_count, if k == split_dimension
        #  - R = rank(operand)
      limit_indices=[l0, l1, ..., lR-1],
        # where
        #   - lk = dim(operand, k) if k != split_dimension
        #   - lk = dim(operand, split_dimension) // split_count, if k == split_dimension
      strides=[t0, t1, ..., tR-1]
        # where
        #   - lk = 0 if k != split_dimension
        #   - lk = dim(operand, split_dimension) // split_count, if k == split_dimension
    ) for j in range(split_count)]
```